File: Arrays/P035.py
from collections import defaultdict

def optimal(nums, k):
    freq = defaultdict(int)
    prefixXor = 0
    count = 0
    for num in nums:
        prefixXor ^= num
        if prefixXor == k:
            count += 1
        if (prefixXor ^ k) in freq:
            count += freq[prefixXor ^ k]
        freq[prefixXor] += 1
    
    return count

nums = [4, 2, 2, 6, 4]
k = 6
print(optimal(nums, k))  

File: Stiver_Problems/Arrays/P035.cpp
// P035.cpp - Count subarrays with given XOR (k)

#include <bits/stdc++.h>
using namespace std;

long long optimal(const vector<int>& nums, int k) {
    unordered_map<int, long long> freq;
    int prefixXor = 0;
    long long count = 0;

    for (int num : nums) {
        prefixXor ^= num;
        if (prefixXor == k) count++;
        int want = prefixXor ^ k;
        if (freq.count(want)) count += freq[want];
        freq[prefixXor]++;
    }
    return count;
}

int main() {
    vector<int> nums = {4, 2, 2, 6, 4};
    int k = 6;
    cout << optimal(nums, k) << "\n";
    return 0;
}
